File: musica/mechanism_configuration.py
"""
Mechanism configuration module for MUSICA.

This module provides classes and functions for creating and configuring
chemical mechanisms for use with MICM solvers.
"""

from ._backend_loader import get_backend

# Get the backend and expose mechanism configuration classes
_backend = get_backend()
_mc = _backend._mechanism_configuration

# Re-export all mechanism configuration classes without underscores
ReactionType = _mc._ReactionType
Species = _mc._Species
Phase = _mc._Phase
ReactionComponent = _mc._ReactionComponent
Arrhenius = _mc._Arrhenius
CondensedPhaseArrhenius = _mc._CondensedPhaseArrhenius
Troe = _mc._Troe
Branched = _mc._Branched
Tunneling = _mc._Tunneling
Surface = _mc._Surface
Photolysis = _mc._Photolysis
CondensedPhasePhotolysis = _mc._CondensedPhasePhotolysis
Emission = _mc._Emission
FirstOrderLoss = _mc._FirstOrderLoss
AqueousEquilibrium = _mc._AqueousEquilibrium
WetDeposition = _mc._WetDeposition
HenrysLaw = _mc._HenrysLaw
SimpolPhaseTransfer = _mc._SimpolPhaseTransfer
UserDefined = _mc._UserDefined
Reactions = _mc._Reactions
ReactionsIterator = _mc._ReactionsIterator
Mechanism = _mc._Mechanism
Version = _mc._Version
Parser = _mc._Parser

__all__ = [
    "ReactionType", "Species", "Phase", "ReactionComponent", "Arrhenius",
    "CondensedPhaseArrhenius", "Troe", "Branched", "Tunneling", "Surface",
    "Photolysis", "CondensedPhasePhotolysis", "Emission", "FirstOrderLoss",
    "AqueousEquilibrium", "WetDeposition", "HenrysLaw", "SimpolPhaseTransfer",
    "UserDefined", "Reactions", "ReactionsIterator", "Mechanism", "Version", "Parser"
]

File: julia/src/mechanism_configuration/mechanism_configuration.jl
# Copyright (C) 2023-2026 University Corporation for Atmospheric Research
# SPDX-License-Identifier: Apache-2.0

"""
    Musica.MechanismConfiguration

Build version 1 mechanism configurations in code and serialize them to JSON or YAML
strings that can be passed to `MICM(config_string = ...)`.

This mirrors the Python (`musica.mechanism_configuration`) and JavaScript
(`musica.mechanismConfiguration`) interfaces.

# Example
```julia
using Musica
using Musica.MechanismConfiguration

A  = Species(name = "A", molecular_weight = 0.029)
B  = Species(name = "B", molecular_weight = 0.029)
AB = Species(name = "AB", molecular_weight = 0.058)

gas = Phase(name = "gas", species = [A, B, AB])

loss = UserDefined(
    name = "AB loss",
    gas_phase = "gas",
    reactants = [ReactionComponent(species_name = "AB")],
    products = [ReactionComponent(species_name = "A"), ReactionComponent(species_name = "B")],
)

mech = Mechanism(name = "ABBA", version = "1.0.0",
                 species = [A, B, AB], phases = [gas], reactions = [loss])

micm = MICM(config_string = to_json_string(mech))
```
"""
module MechanismConfiguration

using JSON
using YAML

include("types.jl")
include("reaction_types.jl")
include("mechanism.jl")

# Core data types
export Species, PhaseSpecies, Phase, ReactionComponent

# Reaction types
export Arrhenius,
    Branched,
    Emission,
    FirstOrderLoss,
    Photolysis,
    Surface,
    TaylorSeries,
    Troe,
    TernaryChemicalActivation,
    Tunneling,
    UserDefined

# Mechanism and serialization
export Mechanism
export to_dict, to_json_string, to_yaml_string, to_string

end # module MechanismConfiguration
